### ISSUE TITLE
We have two copies of ActivatorUtilities

### ISSUE THREAD

**Issue by pranavkm** [aspnet/Common#88]
https://github.com/aspnet/Common/blob/dev/src/Microsoft.Extensions.ActivatorUtilities.Sources/ActivatorUtilities.cs
https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Extensions.DependencyInjection.Abstractions/ActivatorUtilities.cs

They look pretty identical.

**Comment by javiercn** [comment]
@Eilon Do we want to keep the one the first one or the second. My guess is that we want to keep the first one, but it will be good if you can confirm. (By keep I mean the file location, I'll look at the difference in behaviors in both files (if there is any) and change them as appropiate

**Comment by rynowak** [comment]
IMO move the `.Sources` project to the DI repo and link that file into `DI.Abstractions`. The less code in common the better,

**Comment by Tratcher** [comment]
Check the usage. I know the sources are also used here:
https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNetCore.Http.Abstractions/Extensions/UseMiddlewareExtensions.cs#L75

**Comment by javiercn** [comment]
After doing some digging on this one, I've found that the DI version of the file is the most popular one.

Moving everything to use the version on DI has the problem that we don't really want to have a dependency from HttpAbstractions to DI. As far as I understand we don't want to do this because we don't want HttpAbstractions depending on DependencyInjection or viceversa. @Tratcher Could we extract this class into its own library like Microsoft.AspNetCore.MiddleWare (or something like that) and have that take a dependency on DI? In which scenarios are the UseMiddleWareExtensions expected to be used?

```c#
var activator = ApplicationServices.Resolve<IActivator>()
activator.CreateInstance<MyType>(param1, param2, param3)
```
 
Moving everything to use the version in Common implies a big breaking change across many of our repos. According to @pranavkm the DI version also contains tests that act as a specification and that third party library container can use to verify that their containers play nice with ASP.NET expectations.

The solution I propose is keep the DI implementation around as a facade around the implementation in Common, that way we don't have to go an do a big breaking change, we keep a single implementation, and pieces of the framework that don't need or shouldn't take a dependency from DI, can take a source dependency directly from Common.

So to recap:
1) Move UseMiddlewareExtensions to it's own package, take a dependency on DI there, and take a dependency on that package in the places where we use it.
2) Keep the implementation in Common, and make DI a facade that consumes it.

Opinions? @Tratcher @pranavkm @Eilon @rynowak

**Comment by Eilon** [comment]
I think option (2) sounds fine to me. And fairly easy.

**Comment by Tratcher** [comment]
1 isn't an option, more than half the middleware call UseMiddlewareExtensions.

Making Microsoft.AspNetCore.Http.Abstractions depend on Microsoft.Extensions.DependencyInjection.Abstractions wouldn't be the end of the world, but you can resolve that with Fowler who had ActivatorUtilities moved to Common in the first place.

**Comment by javiercn** [comment]
@davidfowl It would imply a big breaking change in all of our repos, that's why it's maintained as a facade over the one in sources, that's what we decided when we visited it. It's not pretty, but it's not the end of the world either.

**Comment by halter73** [comment]
Counter point: The ActivatorUtilities in DI is public, so if we want to remove it, we should probably do it before rtm.

**Comment by javiercn** [comment]
@halter73 That's the thing, we discussed this (see above) and we decided it was too of a breaking change to do at this stage (all other parts of the stack depend on the DI version of it)
